### DISCUSSION TITLE
Problems with predicting minority classes

### DISCUSSION THREAD

**Discussion by trhowell** [ultralytics/yolov5#4734] [Q&A] [answered]
I am training a (pre-trained yolov5m) model on a custom dataset with four classes, but there is a large imbalance between the classes. I am trying to detecting four classes, and each class is present in every image. The primary class typically has 200-300 instances in each image, while the minority classes each have exactly one instance in each image.

In initial training runs, the model is doing well with the majority class I am trying to detect, but it is failing to detect the other three minority classes at all. What are some steps I can take to improve the prediction of the minority classes? I am currently attempting to evolve hyperparameters, but it is a slow process and I'm not sure if there is a more obvious place to start.

One thought I had is to provide training and validation images with only the minority classes present, but this isn't entirely possible as one of those classes is a 'super-class' of the majority class (to give an analogy it's like my model can detect trees, but not forests).

Some additional information:
I am training and detecting with an image size of 3008 pixels (since my majority class can be fairly small) with a batch size of 9 in parallel on 3 GPUs (batch size of 3 is the most that will fit in memory for a single GPU). Other parameters are run at their defaults. I am running 300 epochs but the 'patience' threshold is typically reached around 60-70 epochs. I am slowly adding more labeled data for additional training, but as you can imagine it's slow progress when there are 200-300 objects to label in each image.

Any thoughts, material to read or experiments to try are welcomed!

**Comment by trhowell** [comment, marked as answer]
Ok, so I reduced the image size to 1024 and increased the batch size to 20 and the model is performing MUCH better now and accurately predicting the minority classes.

I'm surprised at the worse performance with the larger image size. The training also ran for more iterations before reaching the patience threshold and the curves in results.png look much cleaner.

If anyone has any insight as to what may have been happening I'd be interested since I'm still learning, but my major roadblock seems to be resolved (for now)!

**Reply by glenn-jocher** [reply]
@trhowell 👋 Hello! Thanks for asking about improving YOLOv5 🚀 training results. 

Most of the time good results can be obtained with no changes to the models or training settings, **provided your dataset is sufficiently large and well labelled**. If at first you don't get good results, there are steps you might be able to take to improve, but we always recommend users **first train with all default settings** before considering any changes. This helps establish a performance baseline and spot areas for improvement.

If you have questions about your training results **we recommend you provide the maximum amount of information possible** if you expect a helpful response, including results plots (train losses, val losses, P, R, mAP), PR curve, confusion matrix, training mosaics, test results and dataset statistics images such as labels.png. All of these are located in your `project/name` directory, typically `yolov5/runs/train/exp`.

We've put together a full guide for users looking to get the best results on their YOLOv5 trainings below.

## Dataset

- **Images per class.** ≥ 1500 images per class recommended
- **Instances per class.** ≥ 10000 instances (labeled objects) per class recommended
- **Image variety.** Must be representative of deployed environment. For real-world use cases we recommend images from different times of day, different seasons, different weather, different lighting, different angles, different sources (scraped online, collected locally, different cameras) etc.
- **Label consistency.** All instances of all classes in all images must be labelled. Partial labelling will not work.
- **Label accuracy.** Labels must closely enclose each object. No space should exist between an object and it's bounding box. No objects should be missing a label.
- **Background images.** Background images are images with no objects that are added to a dataset to reduce False Positives (FP). We recommend about 0-10% background images to help reduce FPs (COCO has 1000 background images for reference, 1% of the total). No labels are required for background images.

<a href="https://arxiv.org/abs/1405.0312"><img width="800" src="https://user-images.githubusercontent.com/26833433/109398377-82b0ac00-78f1-11eb-9c76-cc7820669d0d.png" alt="COCO Analysis"></a>


## Model Selection

Larger models like YOLOv5x and [YOLOv5x6](https://github.com/ultralytics/yolov5/releases/tag/v5.0) will produce better results in nearly all cases, but have more parameters, require more CUDA memory to train, and are slower to run. For **mobile** deployments we recommend YOLOv5s/m, for **cloud** deployments we recommend YOLOv5l/x. See our README [table](https://github.com/ultralytics/yolov5#pretrained-checkpoints) for a full comparison of all models. 

<p align="center"><img width="700" alt="YOLOv5 Models" src="https://github.com/ultralytics/yolov5/releases/download/v1.0/model_comparison.png"></p>

- **Start from Pretrained weights.** Recommended for small to medium sized datasets (i.e. VOC, VisDrone, GlobalWheat). Pass the name of the model to the `--weights` argument. Models download automatically from the [latest YOLOv5 release](https://github.com/ultralytics/yolov5/releases).
```bash
python train.py --data custom.yaml --weights yolov5s.pt
                                             yolov5m.pt
                                             yolov5l.pt
                                             yolov5x.pt
                                             yolov5s6.pt
                                             yolov5m6.pt
                                             yolov5l6.pt
                                             yolov5x6.pt
                                             custom_pretrained.pt
```
- **Start from Scratch.** Recommended for large datasets (i.e. COCO, Objects365, OIv6). Pass the model architecture yaml you are interested in, along with an empty `--weights ''` argument:
```bash
python train.py --data custom.yaml --weights '' --cfg yolov5s.yaml
                                                      yolov5m.yaml
                                                      yolov5l.yaml
                                                      yolov5x.yaml
                                                      yolov5s6.yaml
                                                      yolov5m6.yaml
                                                      yolov5l6.yaml
                                                      yolov5x6.yaml
```

- **P6 models.** P6 models introduced in the [v5.0 release](https://github.com/ultralytics/yolov5/releases/tag/v5.0) such as [YOLOv5s6](https://github.com/ultralytics/yolov5/blob/master/models/hub/yolov5s6.yaml), [YOLOv5m6](https://github.com/ultralytics/yolov5/blob/master/models/hub/yolov5m6.yaml), etc. have a fourth output layer at stride 64 and produce better results for most use cases, particularly with larger images. All P6 models are trained at `--img 1280`. See README [Pretrained Checkpoints](https://github.com/ultralytics/yolov5#pretrained-checkpoints) for additional details.

## Training Settings

Before modifying anything, **first train with default settings to establish a performance baseline**. A full list of train.py settings can be found in the [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py) argparser.

- **Epochs.** Start with 300 epochs. If this overfits early then you can reduce epochs. If overfitting does not occur after 300 epochs, train longer, i.e. 600, 1200 etc epochs.
- **Image size.** COCO trains at native resolution of `--img 640`, though due to the high amount of small objects in the dataset it can benefit from training at higher resolutions such as `--img 1280`. If there are many small objects then custom datasets will benefit from training at native or higher resolution. Best inference results are obtained at the same `--img` as the training was run at, i.e. if you train at `--img 1280` you should also test and detect at `--img 1280`.
- **Batch size.** Use the largest `--batch-size` that your hardware allows for. Small batch sizes produce poor batchnorm statistics and should be avoided.
- **Hyperparameters.** Default hyperparameters are in [hyp.scratch.yaml](https://github.com/ultralytics/yolov5/blob/master/data/hyp.scratch.yaml). We recommend you train with default hyperparameters first before thinking of modifying any. In general, increasing augmentation hyperparameters will reduce and delay overfitting, allowing for longer trainings and higher final mAP. Reduction in loss component gain hyperparameters like `hyp['obj']` will help reduce overfitting in those specific loss components. For an automated method of optimizing these hyperparameters, see our [Hyperparameter Evolution Tutorial](https://github.com/ultralytics/yolov5/issues/607).

## Further Reading

If you'd like to know more a good place to start is Karpathy's 'Recipe for Training Neural Networks', which has great ideas for training that apply broadly across all ML domains:
http://karpathy.github.io/2019/04/25/recipe/
